lib/elixir/unicode/unicode.ex

```diff
@@ -305,70 +331,135 @@ defmodule String.Casing do
                                   to_binary.(title)})
   end
 
-  # Downcase
+  for {codepoint, upper, lower, title} <- codes do
+    def get_codepoint_info(unquote(codepoint)) do
+      if is_nil(unquote(lower)) || unquote(codepoint) == unquote(lower) do
+        %{case: :lower, upper: unquote(upper),
+                        lower: unquote(codepoint),
+                        title: unquote(title)}
+      else
+        %{case: :upper, upper: unquote(codepoint),
+                        lower: unquote(lower),
+                        title: unquote(title)}
+      end
+    end
+  end
+  
+  def get_codepoint_info(_), do: %{}
 
-  def downcase(string), do: downcase(string, "")
+  # Downcase
 
+  def downcase_orig(string), do: downcase_orig(string, "")
+  
   for {codepoint, _upper, lower, _title} <- codes, lower && lower != codepoint do
-    defp downcase(unquote(codepoint) <> rest, acc) do
-      downcase(rest, acc <> unquote(lower))
+    defp downcase_orig(unquote(codepoint) <> rest, acc) do
+      downcase_orig(rest, acc <> unquote(lower))
     end
   end
-
-  defp downcase(<<char, rest::binary>>, acc) do
-    downcase(rest, <<acc::binary, char>>)
+  
+  defp downcase_orig(<<char, rest::binary>>, acc) do
+    downcase_orig(rest, <<acc::binary, char>>)
   end
+  
+  defp downcase_orig("", acc), do: acc
 
+
+  def downcase(string), do: downcase(string, "")
+    
+  defp downcase(<<char :: utf8, rest::binary>>, acc) do
+    ch_down = case get_codepoint_info(<<char :: utf8>>) do
```

Comment:
The Japanese and Korean scripts does not have difference between uppercase and lowercase letters. In the `UnicodeData.txt` file, characters from these scripts has the `Lo` ("letter other") class and the fields `lower`, `upper`, `title` is nil.

I added some text (approximately 300 Kb) in Russian with cyrillic characters. Each cyrillic character  has a size of 2 bytes.

I got the following results. The performance of `downcase` function at 1,8x - 2,2x slower than original version.

I created a branch with name "bench_unicode" in my fork. Steps to reproduce results:
```
git clone -b bench_unicode https://github.com/Chicker/elixir.git
cd elixir/
make clean compile
cd lib/elixir/
../../bin/mix deps.get
../../bin/mix bench --output=""
```